hashes/zkevm/src/keccak/computation_circuit/circuit.rs

```diff
@@ -0,0 +1,343 @@
+    pub(crate) hash_hi: AssignedValue<F>,
+    pub(crate) bytes_left: AssignedValue<F>,
+    pub(crate) word_value: AssignedValue<F>,
+}
+
+impl<F: Field> KeccakComputationCircuit<F> {
+    /// Load keccak assigned rows into halo2-lib.
+    fn load_keccak_assigned_rows(
+        &self,
+        assigned_rows: Vec<KeccakAssignedRow<'_, F>>,
+    ) -> Vec<LoadedKeccakAssignedRow<F>> {
+        let mut loaded_assigned_rows = Vec::with_capacity(assigned_rows.len());
+        let range_circuit_builder = self.range_circuit_builder.borrow();
+        let mut copy_manager = range_circuit_builder.core().copy_manager.lock().unwrap();
+        for assigned_row in assigned_rows {
+            loaded_assigned_rows.push(LoadedKeccakAssignedRow {
+                is_final: copy_manager.load_external_assigned(assigned_row.is_final),
+                hash_lo: copy_manager.load_external_assigned(assigned_row.hash_lo),
+                hash_hi: copy_manager.load_external_assigned(assigned_row.hash_hi),
+                bytes_left: copy_manager.load_external_assigned(assigned_row.bytes_left),
+                word_value: copy_manager.load_external_assigned(assigned_row.word_value),
+            });
+        }
+        loaded_assigned_rows
+    }
+
+    // Generate compact inputs for Poseidon based on raw inputs in Keccak witnesses.
+    //
+    // To illustrate how this function works, let's take the following example:
+    // Parameters: RATE = 2, NUM_WORDS_TO_ABSORB = 2, NUM_ROUNDS = 3
+    // Considering the following logical input in little endian bytes:
+    // logical_input = [0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x21, 0x22, 0x23, 0x24, 0x25]
+    // Keccak witnesses represent the input in 2 chunks(NUM_ROUNDS + 1 rows per chunk):
+    // loaded_assigned_rows = [
+    //     {bytes_left:  21, word_value: 0x0807060504030201, is_final:   N/A},
+    //     {bytes_left:  13, word_value: 0x1817161514131211, is_final:   N/A},
+    //     {bytes_left: N/A, word_value:                N/A, is_final:   N/A},    // No input/output information after first NUM_WORDS_TO_ABSORB rows.
+    //     {bytes_left: N/A, word_value:                N/A, is_final: false},
+    //     {bytes_left:   5, word_value: 0x0000002524232221, is_final:   N/A},
+    //     {bytes_left:   0, word_value: 0x0000000000000000, is_final:   N/A},
+    //     {bytes_left: N/A, word_value:                N/A, is_final:   N/A},    // No input/output information after first NUM_WORDS_TO_ABSORB rows.
+    //     {bytes_left: N/A, word_value:                N/A, is_final:  true},
+    // ]
+    // For easier Poseidon processing, the raw input is encoded in the following way:
+    // If the logic input is empty, [0; RATE].
+    // If the logic input is empty, each word is encoded as RATE witnesses. First word is encoded
+    // as [<byte length of logical input>, logical_input[0], 0..]. Other words are encoded as [logical_input[i], 0, 0..].
```

Comment:
Why are you only putting one `logical_input[i]` in each Poseidon RATE chunk?? Seems very wasteful.

Reply:
Additionally, two keccak words (u64) can be packed together into u128 before doing poseidon.

Reply:
ideally even three u64: 3 * 64 < 254

Reply:
`NUM_WORDS_TO_ABSORB = 17`, can't you pack words into 3*64 bits for 18/3=6 poseidon words, then do it in `RATE=2` chunks for 3 chunks? You can even keep the `length` as a separate value if it's easier.

Reply:
Ah that's much simpler! I thought we cannot pad compact inputs and it makes the code much complicated. Just realize that we have the length in the beginning so it's OK to pad. Nice caught!